TSRM/TSRM.c

```diff
@@ -170,6 +177,11 @@ TSRM_API void tsrm_shutdown(void)
 {
 	int i;
 
+	if (!in_main_thread) {
+		/* ensure singleton */
+		return;
+	}
+
 	if (tsrm_tls_table) {
```

Comment:
I'll check this, thanks for the hint. Thought also to export this as a function, so it's usable in any places with startup/shutdown, maybe improving to explicitly checking thread ids.